src/libexpr/primops/regex.cc

```diff
@@ -0,0 +1,360 @@
+    std::string_view str;
+    uint32_t size_;
+    PCRE2_SIZE* ovector;
+    Regex& re;
+public:
+    MatchData(Regex& re) noexcept
+        : re(re)
+    {
+        match = pcre2_match_data_create_from_pattern(re.code, NULL);
+        size_ = pcre2_get_ovector_count(match);
+        ovector = pcre2_get_ovector_pointer(match);
+    };
+    MatchData(const MatchData&) = delete;
+    MatchData(MatchData&&) = delete;
+    ~MatchData()
+    {
+        pcre2_match_data_free(match);
+    };
+
+    const Regex& regex() const noexcept
+    {
+        return re;
+    };
+
+    uint32_t size() const noexcept
+    {
+        return size_;
+    };
+
+    std::optional<std::string_view> operator[](std::size_t i) const
+    {
+        assert(i < size());
+        if (ovector[2*i] == PCRE2_UNSET) return {};
+        return str.substr(ovector[2*i], ovector[2*i+1] - ovector[2*i]);
+    };
+
+    uint32_t startPos() const noexcept
+    {
+        return pcre2_get_startchar(match);
+    };
+};
+
+int Regex::match(std::string_view str, MatchData& match, PCRE2_SIZE startoffset, uint32_t options)
+{
+    // Cache the string in the match data for match[] to work.
+    match.str = str;
+    // compile if we're using this regex more than once
+    if (this->usage++) this->compile();
+
+    int rc = pcre2_match(code, (const unsigned char*)str.data(), str.length(),
+        startoffset, options, match.match, NULL);
+
+    assert(rc != 0); // match data too small, shouldn't happen
+
+    if (rc < 0 && rc != PCRE2_ERROR_NOMATCH) {
+        unsigned char err[256];
+        pcre2_get_error_message(rc, err, sizeof(err));
+        throw RegexError("unable to match regex: %1%", err);
+    }
+    return rc;
+};
+
+};
+
+struct RegexCache
+{
+    // TODO use C++20 transparent comparison when available
+    std::unordered_map<std::string_view, PCRE::Regex> cache;
+    std::list<std::string> keys;
+
+    PCRE::Regex& get(std::string_view re)
+    {
+        auto it = cache.find(re);
+        if (it != cache.end())
+            return it->second;
+        keys.emplace_back(re);
+        return cache.emplace(keys.back(), keys.back()).first->second;
+    }
+};
+
+std::shared_ptr<RegexCache> makeRegexCache()
+{
+    return std::make_shared<RegexCache>();
+}
+
+size_t regexCacheSize(std::shared_ptr<RegexCache> cache)
+{
+    return cache->keys.size();
+}
+
+static void extract_matches(EvalState & state, const PCRE::MatchData& match, Value & v)
+{
+    auto nameTable = match.regex().nameTable();
+    if (nameTable.size()) {
+        // try to extract named bindings
+        auto bindings = state.buildBindings(nameTable.size());
+        for (auto i : nameTable) {
+            Value & elem = bindings.alloc(i.first);
+            if (!match[i.second].has_value()) elem.mkNull();
+            else elem.mkString(*match[i.second]);
+        }
+        v.mkAttrs(bindings);
+    } else {
+        // the first match is the whole string
+        const size_t len = match.size() - 1;
```

Comment:
it feels like asserting that the size is `> 0` would increase the safety of this function